sklearn/tree/tree.py

```diff
@@ -510,6 +515,111 @@ def decision_path(self, X, check_input=True):
         X = self._validate_X_predict(X, check_input)
         return self.tree_.decision_path(X)
 
+    def prune_tree(self):
+        """Prunes tree using Minimal Cost-Complexity Pruning.
+
+        .. versionadded:: 0.21
+        """
+        check_is_fitted(self, 'tree_')
+
+        if self.alpha == 0.0:
+            return
+
+        n_nodes = self.tree_.node_count
+
+        # prior probability using weighted samples
+        samples = self.tree_.weighted_n_node_samples
+        prior = samples / samples[0]
+
+        # impurity weighted by prior
+        r_node = prior * self.tree_.impurity
+
+        # find parent node ids and leaves
+        child_l = self.tree_.children_left
+        child_r = self.tree_.children_right
+        parents = np.zeros(shape=n_nodes, dtype=np.intp)
+
+        # Uses uint8 to avoid casting when passing to generate_pruned_tree
+        leaves_in_subtree = np.zeros(shape=n_nodes, dtype=np.uint8)
+
+        stack = [(0, -1)]
+        while len(stack) > 0:
+            node_id, parent = stack.pop()
+            parents[node_id] = parent
+            if child_l[node_id] == child_r[node_id]:
+                leaves_in_subtree[node_id] = 1
+            else:
+                stack.append((child_l[node_id], node_id))
+                stack.append((child_r[node_id], node_id))
+
+        # computes number of leaves in all branches and the overall impurity of
+        # the branch. The overall impurity is the sum of r_node in its leaves.
+        n_leaves = np.zeros(shape=n_nodes, dtype=np.int32)
+        leaf_idicies, = np.where(leaves_in_subtree)
+        r_branch = np.zeros(shape=n_nodes, dtype=DTYPE)
+        r_branch[leaf_idicies] = r_node[leaf_idicies]
+
+        # bubble up values to ancestor nodes
+        for idx in leaf_idicies:
+            cur_R = r_node[idx]
+            while idx != 0:
+                par_idx = parents[idx]
+                r_branch[par_idx] += cur_R
+                n_leaves[par_idx] += 1
+                idx = par_idx
+
+        # non-leaves that can be pruned
+        inner_nodes = np.logical_not(leaves_in_subtree)
+        in_subtree = np.ones(shape=n_nodes, dtype=np.bool)
+
+        cur_alpha = 0
+        while cur_alpha < self.alpha:
```

Comment:
2 thoughts:

- on the resources that I read ([here](http://mlwiki.org/index.php/Cost-Complexity_Pruning) and [here](https://newonlinecourses.science.psu.edu/stat508/lesson/11/11.8/11.8.2)), the very first pruning step is to remove all the pure leaves (equivalent to using alpha=0 apparently). This is not done here since `cur_alpha` is immediately overwritten. I wonder if this is done by default in the tree growing algorithm.
- As you check for `cur_alpha < self.alpha` and `cur_alpha` is computed **before** the tree is pruned in the loop, this means that the `alpha` of the returned pruned tree will be greater than `self.alpha`. It would seem more natural to me to return a tree whose `alpha` is less than `self.alpha`. In any case we would need to explain how `alpha` is used in the docs, something like "subtrees whose *scores* are less than `alpha` are discarded. The score is computed as ..."

Reply:
- When alpha = 0, the number of leaves does not contribute to the cost-complexity measure, which I interpreted as "do not prune". Removing the leaves when alpha=0, will increase the cost-complexity measure.

- Returning a tree whose `alpha` is less than `self.alpha` makes sense.

Reply:
>  That is if a node N has 2 child leaves where all the samples in both leaves belong to the same class, then the first step will remove those 2 leaves and make N a leaf

This makes sense. I will review the tree building code to see if this can happen.

To prevent future confusion, I want to get on the same page with our definition of a **pure** leaf. From my understanding, a pure leaf is a leaf whose samples belong to the same class, independent of all other leaves. From reading your response, you consider two leaves to be pure if they are siblings and their samples belong to the same class. Is this correct?

Reply:
>From my understanding, a pure leaf is a leaf whose samples belong to the same class, independent of all other leaves

I meant this as well

Reply:
I just looked at the tree code, I think we can assume that this "first step" is not needed here after all, since a node is made a leaf according to the `min_impurity_decrease` (or deprecated `min_impurity_split`) parameter.

That is, if a node is *pure* according to `min_impurity_decrease`, it will be made a leaf, and thus the example case I mentioned above (a node with 2 pure leaves) cannot exist.